### ISSUE TITLE
🐛 BUG: `wrangler pages dev` messes up redirects and their headers

### ISSUE THREAD

**Issue by newsve** [cloudflare/workers-sdk#5315]
### Which Cloudflare product(s) does this pertain to?

Pages

### What version(s) of the tool(s) are you using?

3.35.0

### What version of Node are you using?

v20.11.1

### What operating system and version are you using?

Ubuntu 22.04.2 LTS

### Observed behavior

When the app redirects wrangler pages dev does the redirect but the URL in the browser's URL bar doesn't change (or redirect's header lacks the redirect status and new URL and the response header lacks a location header.

This is a big problem because it breaks relative links on redirected pages and doesn't reflect  production.

### Expected behavior
A proper redirect with URL change and proper headers

### Steps to reproduce
```bash
git clone https://github.com/newsve/wrangler-no-redirect.git
npm i
npm run dev
```

In your browser go first to `localhost:5713` which is the og SvelteKit/Vite server without wrangler, you see a redirect happng to `localhost:5713/some-random-page`.  Now try the same with `localhost:8788` which is via wrangler, still redirects but the URL in the URL bar does not change.

Also check and compare the headers (left: 5713, right 8788), note that the 8788 one lacks a response location header:

![image](https://github.com/cloudflare/workers-sdk/assets/114379444/47c9641f-7e07-4d86-a2d6-fe57d27f00d0)

### Please provide a link to a minimal reproduction

git clone https://github.com/newsve/wrangler-no-redirect.git

**Comment by GregBrimble** [comment]
Thanks for the report. This can be fixed by building assets to a directory and pointing Wrangler at that (`wrangler pages dev <directory>`), rather than using the proxy feature (`wrangler pages dev -- <command>`). This proxy feature will be deprecated shortly because of incompatibilities like you flag in this issue: #5317.

**Comment by newsve** [comment]
Thanks for your prompt answer.

When I develop I usually don't have a build yet which `wrangler pages dev` could consume. And, I assumed `wrangler pages dev` _is_ for  developing, so running dev envs with HMR and live reload, etc., it is not meant for testing static production builds without hot realod/HMR. Please correct me if I am wrong.

Most modern web frameworks are based on Vite and when I look into, e.g. SvelteKit's output folder `.svelte-kit` I don't see any `cloudflare/` folder when in developing mode (`vite dev' or just 'vite'). Only when I do a build with 'vite build`, then I see that folder. But actually I never do latter because CF does this automatically when pushing to GH. So, not sure how the dev workflow would be then.

So, maybe I do miss something and you can hint me with an example wrangler command targeting a Vite framework such as SvelteKit how this should work when developing not. Thanks!

**Comment by newsve** [comment]
But will wrangler pages dev still inject Cloudflare binds, e.g. to services like KV, or the props on `request.cf`, so the framework can consume those? This is somewhat the key USP of wrangler.

**Comment by GregBrimble** [comment]
Yes, it will.

**Comment by newsve** [comment]
Just an idea:

Right now, I use only the output of vite's dev server on 5173 despite wrangler started and proxied latter. And it is great because it has already all injected binds without the need to use wrangler's proxied version on 8788.

A dev doesn't need more. So if there is way to get just the binds injected into vite dev without actually wrangler proxying vite would be already a huge step (because something like vite build --watch isn't as nice as vite dev for development).

**Comment by IgorMinar** [comment]
hi @newsve, if you use the latest cloudflare adapter for sveltekit it will allow you to access all the bindings from within sveltekit's dev server. you can scaffold a new project using create-cloudflare to test it out and check out https://developers.cloudflare.com/pages/framework-guides/deploy-a-svelte-site/#sveltekit-cloudflare-configuration - the docs don't mention it yet (WIP) but you can configure the bindings via wrangler.toml. Again, check out a clean create-cloudflare project where all of this is preconfigured already.

**Comment by newsve** [comment]
Thanks, just skimmed through a create-cloudflare installation, and wow,l super impressive! All the supported framework in one place, amazing!

One question though: I can't use the locally emulated Cloudflare services anymore like with wrangler in dev mode (wrangler doesn't run with `npm run dev`, just with `build` and `preview`), so I would need to bind my local dev server with an actual Cloudflare service in the cloud and have there always min. two of each kind (one for dev and one for production).

Did I get it right?

**Comment by IgorMinar** [comment]
> Did I get it right?

not quite, with the create-cloudflare integration we start a wrangler/miniflare process behind the scenes and integrate it into vite. So you can access a local D1 database, or KV store from your app while it's running via `vite dev`

**Comment by newsve** [comment]
Incredible! Is vite then somehow patched? Because in package.json, I only find `"dev": "vite dev",` and in vite.config.js I find no invocation of wrangler/miniflare either:
```js
import { sveltekit } from '@sveltejs/kit/vite';
import { defineConfig } from 'vitest/config';

export default defineConfig({
        plugins: [sveltekit()],
        test: {
                include: ['src/**/*.{test,spec}.{js,ts}']
        }
});
```

**Comment by newsve** [comment]
tl;dr:

- Deprecating Wrangler's proxy functionality risks degrading developer experience by losing key development features like live reload and HMR, repro steps for Svelte below but this issue is with every framework, Svelte acts as an example her
- This change may push developers towards other platforms due to reduced testing convenience; this is critical because the workers-sdk acts as the gateway and catalyst for utilizing all other Cloudflare services
- Suggestion: Allow Cloudflare feature injection in npm run dev without full proxy use, to retain essential development tools.

Hey Cloudflare Team,

Wanted to flag again a potential issue with deprecating Wrangler's Pages Dev Proxy functionality. Please follow these steps setting up a Svelte project with your new create-cloudflare CLI (which is again highly impressive!):
```bash
npm create cloudflare@latest // Chose Create aoo/Website o. app/Svelte/Typescript
```
Now. add `src/routes/+page.server.ts`: 
```ts
import type { PageServerLoad } from './$types'

export const load: PageServerLoad = async ({ request }) => {
  console.log(request.cf)
}

```
Try `npm run dev` and check the server logs, access the new app in your browser, console.log prints nothing.

Try `npm run preview` and chack the server logs, and yes, `request.cf` is set! but you don't get:
1. __Must-have:__ Rebuild on file changed (somehow doesn't work)
2. __Must-have:__ Live-reload in the browser (this wouldn't work if even 1 did work)
3. __Nice-to-have:__ HMR in the browser (this wouldn't work if even 1 did work)

2 and 3 rely on Vite's unmatched dev server which is key to Vite's success and which you can't use with `npm run preview`

The current developer experience (DX) with Cloudflare services takes a hit due to the necessity of restarting the server or refreshing the browser on every file change. This becomes especially problematic for those relying on Cloudflare-specific features, such as request.cf, to dynamically adjust behavior based on the node's location. The cumbersome workaround of using `npm run preview` isn't a viable solution for a smooth and efficient development workflow.

I'm concerned this move might push developers towards alternative solutions or other providers, particularly when the ease of integration and testing of CF services is compromised. The deprecation of wrangler pages dev -- could lead to missed opportunities in leveraging Cloudflare to its full potential.

Would it be possible to explore a solution that allows injecting Cloudflare's features into the development environment, specifically through `npm run dev`, without fully acting as a proxy? Currently, I bypass the proxy's verbosity on 8788 and limitations by directly using Vite's output on 5173. So, we would need just the injecting part of `wrangler pages dev ++ vite`.

**Comment by aroman** [comment]
Strongly agree with @newsve's points above... I was shocked to learn that wrangler is deprecating the proxy feature without a viable replacement (`vite build --watch` will not meet our needs). 

To be clear: hot-reload support is not a "nice-to-have" for us, it is **essential** to our workflow: we literally cannot develop without it. Our application has a lot of intricate client state (we're building games) and forcing hard-reloads fundamentally breaks our team's workflow. cc: @admah

**Comment by newsve** [comment]
fwiw, I changed issue title/subject since the actual issue evolved to env var injection to hot reloading dev servers such as `vite dev` (instead of starting a new thread), hope that's ok and look forward to CF's thoughts and ideas on how to tackle this issue

**Comment by malobre** [comment]
And even if you don't need HMR, the live-reload feature does not work, see #5351.

**Comment by ducan-ne** [comment]
Try to upgrade to the latest of wrangler and then need to downgrade because of this, at least provide a way to make it work with current tooling, AFAIK it's impossible to do it with Vite to support HMR, bundleless, etc. Or you guys want me to move to rollup and rollup --watch

**Comment by ducan-ne** [comment]
As user perpective, I think honox API is the BOAT here, they provide a vite plugin and I don't need to do anything
Below is my command, is pretty long but runable
```
pnpm dlx kill-port-process 8788 && wrangler pages dev --compatibility-date=2024-04-01 --compatibility-flag nodejs_compat --proxy 5173 --log-level log -- vite --host 0.0.0.0 -l silent
```
// ps: yes I need to `kill-port-process 8788` because wrangler randomly not teardown process if it's crash or something
honox is pretty good at DX, however they generate a _worker.js as output, which is I understand will cost 1 request for each static file.

**Comment by jonaswre** [comment]
After chasing this problem for 2 days. I realised there is a way to integrate miniflare without using wrangles pages dev at all. 

the wrangler package provides a function called getPlatformProxy that will give you access to everything. 

The svelte adapter handles it in the following manner.

https://github.com/sveltejs/kit/blob/bbab296f6fcc05af6b999182798bcdedabbaa4c9/packages/adapter-cloudflare/index.js#L130

**Comment by AVGVSTVS96** [comment]
**+1**

I'm trying to find a solution to be able to run a dev server for my Astro + Pages Function GPT Chat application without `-- <command>`. Like mentioned by others here, it's important to be able to use live reloading when making changes in a dev environment without have to rebuild and restart the dev server every time.

I wrote a discussion post on this with more details about my workflow and project: [How do I start an Astro + Pages Function dev server if -- <command> is being deprecated?](https://github.com/cloudflare/workers-sdk/discussions/5617)

---

@jonaswre Could you elaborate on this? I'm a new programmer and am struggling to understand how to implement this solution for my Astro project.

> After chasing this problem for 2 days. I realised there is a way to integrate miniflare without using wrangles pages dev at all.
> 
> the wrangler package provides a function called getPlatformProxy that will give you access to everything.
> 
> The svelte adapter handles it in the following manner.
> 
> https://github.com/sveltejs/kit/blob/bbab296f6fcc05af6b999182798bcdedabbaa4c9/packages/adapter-cloudflare/index.js#L130

**Comment by jonaswre** [comment]
@AVGVSTVS96 seems like astro has the same concept like sveltekit. But from the doc it seems like its not enabled by default.

https://docs.astro.build/en/guides/integrations-guide/cloudflare/#platformproxy

This should help

**Comment by newsve** [comment]
@jonaswre Thanks for looking into it!

> I realised there is a way to integrate miniflare without using wrangles pages dev at all.
> the wrangler package provides a function called getPlatformProxy that will give you access to everything
> The svelte adapter handles it in the following manner.

Please check again my mini example which indeed uses the cloudflare-adapter in  https://github.com/cloudflare/workers-sdk/issues/5315#issuecomment-2012389845

But while using the cloudflare-adpater it still doesn't exposes anything in dev mode.

>     "preview": "npm run build && wrangler pages dev .svelte-kit/cloudflare",
> This is a way to have a look at something that closely resembles how its running in prod.

`npm run preview` is amazing but devs need `dev` mode for a close feedback loop with all the thing which preview lacks: rebuild on file change, live reload and/or HMR in the browser after rebuild and all in few ms. `preview` doesn't give us any of that.

How do you want to develop against all the services such as D1, KV, DO, etc.? Using native binds doesn't work in `dev` mode and `preview` is way too clunky to do anything more than a hello world or do you want to manually rebuiild, wait and refresh your browser on any little change? So you go back to use REST calls with API tokens in env vars just like before. But then why do we need wrangler in the first place (except for managing the cloud resources in a dashboard manner once or twice a week) ?

**Comment by jonaswre** [comment]
Thats the Point.
With the Platform Proxy you should just be able to use the normal dev server from your framework. No need for wrangler dev. 

In sveltekit I can just use "vite dev" as usual the platform proxy injects everything it should. 

Have you enabled the platform proxy as seen in the like I provided? 

If you need further help please provide your Astro config where you configure the adapter. 

One more thing to notice. You need to use .dev.vars instead/in combination with your .env file. Miniflare will use envs from .dev.vars instead of .enc.

**Comment by KyGuy2002** [comment]
I am having the same problem.  I understand the reasoning behind wanting to switch away from this, but if a viable alternative is not available then removal is not an option.

Reading the blog and other issues, it looks like CF is working with the full stack frameworks to make it integrate nicely with them, but for plain Vite or any other unsupported tools, there is no way to do this anymore.

I think this should be added back with a disclaimer about the potential issues, but still allow its use for projects that rely on it.

**Comment by jonaswre** [comment]
@KyGuy2002

There is.

The wrangler npm package exposes a getPlatformProxy function. 

You can use it in the same way that the svelte or astro adapter use it. 

If you look at the source code of those adapters, you will see they don't use any specific tools for the integration that you won't be able to use in "plain" vite. 

Basic concept is: 

let DB = env.db

If(env == Dev) {
   platform = getPlatformProxy()
   DB = platform.env.db
}

DB.query(...)


The frameworm adapters inject this, in the way they think platform specifics should be handled. 

For svelte in all functions that could interact with the platform you have access to a event object. That event object has a platform property which always contain the bindings from wrangler. In Prod/Preview they contain the real deal. In dev they referenced the local services. 

I think the development workflow is meant to be the following:


1. Dev your features using your hmr development server ( access cloudflare using the getPlatformProxy)
2. Build
3. Verify your result using a better representation of the cloudflare infrastructure using "wrangler dev <build_path>"
4. Deploy to Preview 
5. Deploy to Production

**Comment by KyGuy2002** [comment]
@jonaswre 

Thanks for responding.

I'm not sure im understanding what you mean.  I am using the vanilla pages functions in the `functions` directory, and then my react site from `src`.  Don't I need `vite dev` to serve the react app, and wrangler proxy those requests and handle the functions ones?

Thanks.

**Comment by AVGVSTVS96** [comment]
@jonaswre If I understand correctly, I should be able to enable platformProxy for cloudflare adapter in astro config file and cloudflare should inject bindings into my dev server? Does bindings include ability to access my cloudflare function, or is this just for database stuff?

I tried this and have not been successful, no access to function in dev mode.

Do I also have to setup something else to access the cloudflare runtime? Or should this just work?

Here's a link to the branch in my repo where I tried this: https://github.com/AVGVSTVS96/astroSite/tree/cloudflareProxy

**Comment by newsve** [comment]
> 1. Dev your features using your hmr development server ( access cloudflare using the getPlatformProxy)

@jonaswre would you mind to make a small poc repo? Even C3 (`npm create cloudflare@latest`) with SvelteKit as app doesn't expose anything nor does it run wrangler dev along vite. Or just take @AVGVSTVS96 repo and add the missing parts.

Without being able to access the native, og CF bindings in a hot-reloading dev environment, how should we develop against killer products such as CF KV?

**Comment by jonaswre** [comment]
https://github.com/jonaswre/super-math-5233

If created a small example that does something useless with KV. 

Steps do reproduce:

1. npm create cloudflare@latest
2. chose svelte
3. wrangler kv:namespace create wrangler_demo   
4. add config to wrangler.toml
5. Add types to app.d.ts
6. Use KV in hooks
7. npm run dev


But I guess the "magic" is how to access the KV. Don't use process.env but event.platform.env.

src/hooks.server.ts
```typescript
import type { Handle } from "@sveltejs/kit";

export const handle: Handle = async ({ event, resolve }) => {
	if(event.platform !== undefined) {
		const key = "SET_ME";
		const last_value =  await event.platform.env.wrangler_demo.get(key);
		const new_value = (last_value ?? '')+1;
		await event.platform.env.wrangler_demo.put(key, new_value);
		event.locals.current = new_value;

	}
	return await resolve(event);
}
```

src/app.d.ts
```typescript
// See https://kit.svelte.dev/docs/types#app
// for information about these interfaces
declare global {
	namespace App {
        interface Locals {
            current: string | null;
        }
        interface Platform {
            env: {
				wrangler_demo: KVNamespace;
			};
            cf: CfProperties
            ctx: ExecutionContext
        }
    }
}

export {};
```

**Comment by AVGVSTVS96** [comment]
@jonaswre Is this not possible in my static Astro project?

> @jonaswre If I understand correctly, I should be able to enable platformProxy for cloudflare adapter in astro config file and cloudflare should inject bindings into my dev server? Does bindings include ability to access my cloudflare function, or is this just for database stuff?
> 
> I tried this and have not been successful, no access to function in dev mode.
> 
> Do I also have to setup something else to access the cloudflare runtime? Or should this just work?
> 
> Here's a link to the branch in my repo where I tried this: https://github.com/AVGVSTVS96/astroSite/tree/cloudflareProxy

**Comment by jonaswre** [comment]
I have no experience with astro but from the documentation you should be able to. 

https://docs.astro.build/en/guides/integrations-guide/cloudflare/#cloudflare-runtime

```
const { env } = Astro.locals.runtime;
```

It seems like the folloing line injects the platformProxy the same way its handled in svelte. 
https://github.com/withastro/adapters/blob/38c32c9db490dbe58d813c7eed1113d4f7f8683d/packages/cloudflare/src/index.ts#L194

**Comment by rschristian** [comment]
> You can use it in the same way that the svelte or astro adapter use it.
> 
> If you look at the source code of those adapters, you will see they don't use any specific tools for the integration that you won't be able to use in "plain" vite.

If this is going to be the recommended path forward, the community really needs a spec implementation rather than piecing together what some frameworks have thrown together (not to mention that has no guarantee of being correct).

**Comment by jonaswre** [comment]
> If this is going to be the recommended path forward, the community really needs a spec implementation rather than piecing together what some frameworks have thrown together (not to mention that has no guarantee of being correct).

I think it would be best for the deprecation warning to guide people to the following page: https://developers.cloudflare.com/workers/wrangler/api/#getplatformproxy and then add a section that lets people know that if they use a framework they should check the adapters documentations. 
If you have some very custom setup you need to put some thought into your dev mode anyway.

I think the information is all there they should just point people to the new solution. Just telling people it will no longer work creates threads like this.  

And if there are a lot of people, just using vite without any other framework, they could come together an implement a vite plugin that handles it for vite. 

As I said not much magic going on in the adapters. 

## [Astro](https://github.com/withastro/adapters/blob/f14f4d72befcb715da806fea54a8245e8f052556/packages/cloudflare/src/index.ts#L188)
```
const platformProxy = await getPlatformProxy({
		configPath: args.platformProxy.configPath ?? 'wrangler.toml',
		experimentalJsonConfig: args.platformProxy.experimentalJsonConfig ?? false,
		persist: args.platformProxy.persist ?? true,
	});

	const clientLocalsSymbol = Symbol.for('astro.locals');

	server.middlewares.use(async function middleware(req, res, next) {
		Reflect.set(req, clientLocalsSymbol, {
			runtime: {
				env: platformProxy.env,
				cf: platformProxy.cf,
				caches: platformProxy.caches,
				ctx: {
					waitUntil: (promise: Promise<any>) => platformProxy.ctx.waitUntil(promise),
					passThroughOnException: () => platformProxy.ctx.passThroughOnException(),
				},
			},
		});
		next();
	});
```
1. call platform proxy
2. inject in the way they handle platform specifics. 

## [Sveltekit](https://github.com/sveltejs/kit/blob/7a625724e240c05f3b1b1b000bc5f51333982522/packages/adapter-cloudflare/index.js#L131)
```
const proxy = await getPlatformProxy(options.platformProxy);
	const platform = /** @type {App.Platform} */ ({
		env: proxy.env,
		context: proxy.ctx,
		caches: proxy.caches,
		cf: proxy.cf
	});

	/** @type {Record<string, any>} */
	const env = {};
	const prerender_platform = /** @type {App.Platform} */ (/** @type {unknown} */ ({ env }));

	for (const key in proxy.env) {
		Object.defineProperty(env, key, {
			get: () => {
				throw new Error(`Cannot access platform.env.${key} in a prerenderable route`);
			}
		});
	}

	return {
		platform: ({ prerender }) => {
			return prerender ? prerender_platform : platform;
		}
	};
```
1. call platform proxy
2. inject in the way they handle platform specifics. 

## [Remix](https://github.com/remix-run/remix/blob/b3f75da44749d6d69dd613668fd346b0e6802d9c/packages/remix-dev/vite/cloudflare-proxy-plugin.ts#L60)
```
 let { getPlatformProxy } = await importWrangler();
      // Do not include `dispose` in Cloudflare context
      let { dispose, ...cloudflare } = await getPlatformProxy<Env, Cf>(options);
      let context = { cloudflare };
      return () => {
        if (!viteDevServer.config.server.middlewareMode) {
          viteDevServer.middlewares.use(async (nodeReq, nodeRes, next) => {
            try {
              let build = (await viteDevServer.ssrLoadModule(
                serverBuildId
              )) as ServerBuild;

              let handler = createRequestHandler(build, "development");
              let req = fromNodeRequest(nodeReq);
              let loadContext = getLoadContext
                ? await getLoadContext({ request: req, context })
                : context;
              let res = await handler(req, loadContext);
              await toNodeRequest(res, nodeRes);
            } catch (error) {
              next(error);
            }
          });
        }
      };
```
1. call platform proxy
2. inject in the way they handle platform specifics. 

While looking through the implementation it would probably be very easy to change the remix adapter to a standalone vite adapter.

https://github.com/remix-run/remix/blob/b3f75da44749d6d69dd613668fd346b0e6802d9c/packages/remix-dev/vite/cloudflare-proxy-plugin.ts#L34

**Comment by rschristian** [comment]
> if they use a framework they should check the adapters documentations. If you have some very custom setup you need to put some thought into your dev mode anyway.

There's a huge gap between not using a framework and "very custom setup" -- a plain Vite/Webpack/Parcel app is not a "very custom setup".

Pointing at what frameworks have cobbled together, again, is not sufficient. A spec (frameworkless, vanilla) implementation is the bare minimum needed to have a healthy migration, else, people will stick to old versions.

**Comment by jonaswre** [comment]
You are write having to look for the solution is annoying. That's why I suggested to add a reference to the docs. 

But they already did a pretty nice job of hiding the details. 

```
export let kv = process.env.kv1;
If (process.env.DEV == true) {
   const {platform} = getPlatformProxy();
   kv = platform.env.kv1
}
```

This is pretty great to emulate the whole cloudflare infrastructure. 

You haven't seen bad dx if you didn't try to develop web services in azure. 

In open source land there is always the option to develop a vite/webpack/parcel plugin. 

The reason why I think plain JavaScript is custom because most people who develop web apps usually pick framework or separate server and client code. And in the regular client spa you can't access cloudflare services directly anyways.

**Comment by rschristian** [comment]
> You are right having to look for the solution is annoying. That's why I suggested to add a reference to the docs.

I, and I'd imagine all others, have read the docs. While the `getPlatformProxy` API is indeed loosely documented, emulating the full CloudFlare suite is an exercise entirely left up to the reader. This is not good.

> This is pretty great to emulate the whole cloudflare infrastructure.
>...
> In open source land there is always the option to develop a vite/webpack/parcel plugin.

Neat, where's the docs showing off integrations with the full suite? How to manage the dichotomy of a dev mode that is entirely built off of custom, third-party middlewares while in prod we have rigid file system-based routing requirements and a marginally different API?

> And in the regular client spa you can't access cloudflare services directly anyways.

I think you're missing the forest for the trees here. You can access these services via a worker/page function/etc., which in dev the proxy is a rather good option for, hence why we're here.

**Comment by bjankord** [comment]
I was a bit disappointed to learn the `proxy` flag was being deprecated. It made building a SPA + CF Functions integrated together in 1 repo so easy.  

I understand you could get something similar working with running `vite build --watch` and simultaneously `wrangler pages dev <directory>`. however, this comes with some cons:
- Using `vite build --watch` makes it harder to debug your app code as its compiled with build settings (minified code/split into chunks)
- You miss out on HMR, you need to manually refresh pages whenever you want to see updates (not having HMR after using it for years is a painful developer experience)
- You can't easily use tools like React Dev Tools / Vue Dev Tools, these work best when your app is built in dev mode. (there are some custom extensions you can force these to work but the experience is not great)

I'm also familiar with various frameworks adding `getPlatformProxy` support. I'd prefer not to use svelte, astro, or remix, when I'm building simple app (TODO) and just stick with a SPA app + CF Functions (part of the nicety of CF Pages is simplicity).

For developers like me, I think it would be nice if Cloudflare added some examples of how to integrate 
 `getPlatformProxy` support in a basic react app with vite (no remix, astro, next.js, etc).  Even better if CF provided a npm package for this but maybe that's asking too much.

Alternatively, I also understand that @IgorMinar has been working with the Vite team on the new Environments API that looks like it may solve some of these concerns I, and other developers, have around trying to build simple SPAs on CF Pages with HMR support while running `wrangler pages dev`.

Similar to `getPlatformProxy`, I think once Vite 6 is release and the Environment API is stable, it would be great for Cloudflare to document some basic examples of how to use these to integrate CF bindings during local development + enable HMR.

**Comment by connorads** [comment]
💡  TIL: `vite dev` handles local CF/miniflare/KV for you automagically if you're using SvelteKit (and other adapters)
https://blog.cloudflare.com/blazing-fast-development-with-full-stack-frameworks-and-cloudflare/

I used the `-- proxy` option a [2 years ago](https://github.com/connorads/connoradams.co.uk/commit/5f47b79f5f1679a87d1a11c95202f58f7c73fff8) and thought I needed it for SvelteKit now. But I don't.
I assumed `npm run dev` (aka `vite dev`) couldn't possibly be running the Wrangler KV in the background, but it is. And HMR works too.
Thought I would just comment here in case anyone else is in a similar confused state trying to run a non-SPA app locally with CF KV etc.

**Comment by d3v1an7** [comment]
11ty user here, moving across from `netlify dev`, which allows you to define a build command in the `.toml`.

The cludgy workaround for now is:
```
@11ty/eleventy --watch && npx wrangler pages dev _site --live-reload
```

Then nudging `wrangler.toml` when the build finishes:
```
  eleventyConfig.on('eleventy.after', async () => {
    const now = new Date();
    fs.utimes('./wrangler.toml', now, now);
  });
```

It works, but I don't love it!

Was surprised that `wrangler.toml` didn't seem to support the build command (unless I've missed something?), especially seeing as it can be set via the UI.

**Comment by vmuthal** [comment]
I don't know why Cloudflare team decided to deprecate the -- command or proxy feature. 
For whatever reasons they have done this it just put developers into very much confusion about how to run the Vite Server with Cloudflare Pages ie Wrangler.

Problems: 
- Now what to do when it got deprecated and difficult to understand further solutions given by others. 
- How to run both vite and cloudflare pages workers functions simultaneously?
- We can't use vite build --watch (Not Good Solution)

My Analysis and Solution:
- We want to run Vite Sever
- We want to run Wrangler Dev
- We need HMR

How to do it? 
So as of now I am using Wrangler Version 3.99. This version of Wrangler supports --proxy.  And with Concurrently package we can run the commands simultaneously. So Lets run vite and wrangler simultaneously and use the proxy feature of wrangler.

So earlier it was:
Wrangler Version Less than < 3
```
"scripts": {
    "dev": "wrangler pages dev --port 8789 -- npx vite -- proxy 5174",
}
```

And now it is:
Wrangler Version Greater than > 3
```
"scripts":{
	"dev": "concurrently --kill-others-on-fail \"pnpm run dev:vite\" \"pnpm run dev:wrangler\"", 
	"dev:vite": "vite", 
	"dev:wrangler": "wrangler pages dev --proxy 5174 --port 8789",
}
```

This is just running two servers simultaneously and making things easier to understand. In earlier version of wrangler they had support for command as well as proxy as an arguments. Now they only support proxy as an argument. So using concurrently I am able to run both properly without much of difficulty to understand.

This comment is just to help others so that things can stay simple and also it is working now and I don't know if in future they remove --proxy argument. Then things will fail.

**Comment by ilyakam** [comment]
@vmuthal solution also works with Astro! Here's how:

1. [Set up](https://developers.cloudflare.com/pages/framework-guides/deploy-an-astro-site/#set-up-a-new-project) a new Cloudflare Pages + Astro project
2. `pnpm add -D concurrently`
3. Update the `"scripts"` in `package.json`:

```json
{
  "scripts": {
    "dev": "concurrently --kill-others-on-fail \"pnpm run dev:astro\" \"pnpm run dev:wrangler\"",
    "dev:astro": "astro dev",
    "dev:wrangler": "wrangler pages dev --proxy 4321 --port 8788"
  }
}
```

---

To verify the KV bindings:

1. Add your `kv_namespaces` to `wrangler.json` per [these docs](https://developers.cloudflare.com/workers/wrangler/configuration/#sample-wrangler-configuration)

2. Update `Welcome.astro` to:

```astro
---
const { env } = Astro.locals.runtime;

const foo = await env.FOO.list(); // or `.get()` etc.
---
```

**Comment by madsem** [comment]
Not a solution @vmuthal as they are removing the `--proxy` flag... 
That will make it impossible to work in monorepos / multi site repos btw @GregBrimble 

Would require largescale changes, building / copying builds into directories to serve from, not to mention CI pipelines.
I'm sure there are good reasons to want to remove this, but personally I'm pretty frustrated by this and the lack of alternatives.

"Build into a directory and start wrangler from there" is not a real alternative...

**Comment by ricky11** [comment]
Cant work with pages functions without HMR in dev mode, is their a solution?

**Comment by braden-w** [comment]
As a temporary fix, I ended up with this hybrid approach.

```typescript
import type { D1Database } from "@cloudflare/workers-types";
import * as schema from "@repo/db";
import { getEvent } from "@tanstack/react-start/server";
import { drizzle } from "drizzle-orm/d1";

type Env = { DB: D1Database };

async function getEnv(): Promise<Env> {
	// In development, we can use getPlatformProxy (https://developers.cloudflare.com/workers/wrangler/api/#getplatformproxy)
	// We cannot use it in production because `import { getPlatformProxy } from "wrangler"` causes an error in build
	if (process.env.NODE_ENV === "development") {
		const { getPlatformProxy } = await import("wrangler");
		const { env } = await getPlatformProxy<Env>();
		return env;
	}
	// In production, the bindings are attached to the event in event.context.cloudflare.env https://nitro.build/deploy/providers/cloudflare#direct-access-to-cloudflare-bindings
	const env = getEvent().context?.cloudflare?.env as Env | undefined;
	if (!env) {
		throw new Error("Cloudflare environment is not attached to the event.");
	}
	return env;
}

// Example creating a context for a TRPC router
export async function createContext() {
	const env = await getEnv();
	const db = drizzle(env.DB, { schema, logger: true });

	return { db };
}
```

1. **In Dev Mode:** Use `const { env } = await getPlatformProxy<Env>()`. `getPlatformProxy` creates an object that emulates the Cloudflare bindings in Node.js, as explained [[here](https://developers.cloudflare.com/workers/wrangler/api/#getplatformproxy)](https://developers.cloudflare.com/workers/wrangler/api/#getplatformproxy).

   - Dynamically imported with `const { getPlatformProxy } = await import("wrangler")` to prevent wrangler from being included in production build. This is because although it works in dev, it doesn't work with `vinxi build` in TanStack Start because the `import { getPlatformProxy } from "wrangler"` gives the following error:

     > ... ERROR Unexpected token production. Expected identifier in /Users/.../.../.../node_modules/.pnpm/wrangler@4.7.0\_@cloudflare+workers-types@4.20250404.0/node_modules/wrangler/wrangler-dist/cli.js

     ![Image](https://github.com/user-attachments/assets/14d4b29b-2d56-47d0-a479-274fa1161c45)

2. **In Production:** Use `const event = await getEvent()` (or your framework's equivalent) to access the event in Nitro.

   - In Cloudflare, Nitro attaches the Cloudflare bindings to `event.context.cloudflare.env` , as described [[here](https://nitro.build/deploy/providers/cloudflare#direct-access-to-cloudflare-bindings)](https://nitro.build/deploy/providers/cloudflare#direct-access-to-cloudflare-bindings).


Now, I can access Bindings in `vinxi dev` with HMR rather than having to use `wrangler pages dev`. This is still a hackery fix, hopefully not necessary as bundling in dev and production gets more unified.

**Comment by braden-w** [comment]
Update, Cloudflare has recently released [announced](https://blog.cloudflare.com/introducing-the-cloudflare-vite-plugin/#workers) their official Vite plugin (`@cloudflare/vite-plugin`) that can integrate with any Vite-based application and emulate the worker runtime in development.

You can now simply:

```typescript
// vite.config.ts
import { defineConfig } from "vite";
import { cloudflare } from "@cloudflare/vite-plugin";

export default defineConfig({
  plugins: [cloudflare()]
});
```

in any Vite app.

Now, our Worker code can run inside the native Cloudflare Workers runtime (workerd) during development, and the dev server matches the production behavior as closely as possible (compared to the `getPlatformProxy` approach from before).

**Comment by KyGuy2002** [comment]
Has anyone been able to get the new vite plugin to work with cloudflare pages?

For me it only serves the static assets, I think its just treating it as a worker and not a pages project so its ignoring the `/functions` directory.

**Comment by anhtuank7c** [comment]
Has anyone successfully run a **SvelteKit** application with `@cloudflare/vite-plugin`?
I tried to edit `vite.config.ts` as @braden-w mention from above but doesn't seem to work.
I use workers to run SvelteKit application, NOT pages.

https://developers.cloudflare.com/workers/static-assets/migration-guides/migrate-from-pages/

**Comment by penalosa** [comment]
As mentioned earlier in this thread, `@cloudflare/vite-plugin` is the recommended approach for local dev with Vite and Cloudflare bindings. It runs your Worker in the native workerd runtime with full HMR support, and so no proxy layer is needed.

We're gradually expanding framework support. Today, React Router, TanStack Start, and Astro integrate with the Vite plugin directly, and more frameworks are being added over time.

For frameworks not yet supported, [`getPlatformProxy()`](https://developers.cloudflare.com/workers/wrangler/api/#getplatformproxy) remains available.
